### PR TITLE
Add vision for Gemma3

### PR DESCRIPTION
**Pull request by abheesht17**
Bunch of notebooks to demonstrate how different components work:

`CausalLMPreprocessor`: https://colab.research.google.com/drive/1fAb3Rvrw2zRZd5gfVCmB5WW2qdWl_eMN?resourcekey=0-uJD6GoFVgREFVSeMVaXAGA&usp=sharing

`.generate()`: https://colab.research.google.com/drive/1l2atV5VNt9HYKk-BZ39YcQBcsNG4D94V?resourcekey=0-qX4Xka61fikIygXpFZ-eGA&usp=sharing

`.fit()` on a randomly initialised model (because even the 4B one cannot fit on A100): https://colab.research.google.com/drive/11Yi9oAtBs9VvrwJPmtY9bsr8ZajwRhl6?resourcekey=0-wKYPYAm9uw51k1T39DfOGg&usp=sharing

JAX

`generate()`: https://colab.research.google.com/drive/1AR7G3UFawqO9VsNDTtLUrJP-dRTvbkh1?resourcekey=0-4ofXHp_ko1LDzFIG9pYkdQ&usp=sharing

`.fit()`: https://colab.research.google.com/drive/15IF3wF55EIbuenRB99R_JgbYI8Z_eA9U?resourcekey=0-w0sOujM6cxO1qsvkTXYiDA&usp=sharing

### CODE REVIEW

**Comment by abheesht17** [comment]
Okay! @mattdangerw - this is ready for review, mostly. I'm filling up the doc-strings and the unit tests, but can review the rest.

Also, probably a good idea to refer to this, because a lot of the vision components were added in the previous PR: https://github.com/keras-team/keras-hub/pull/2152.

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 1]
```diff
@@ -273,162 +371,151 @@ def call(
...
-            tf.math.reduce_sum(
-                tf.cast(
-                    tf.math.greater(
-                        num_valid_images, self.max_images_per_prompt
-                    ),
-                    dtype=tf.int32,
+        # == Branch: vision model, with non-`None` value for `images` ==
+
+        # Check: token IDs should not have less than 1, or more than
+        # `max_images_per_prompt` start of image tokens.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:424]: remove commented code

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 2]
```diff
@@ -489,174 +606,162 @@ def generate_preprocess(
...
-            tf.math.reduce_sum(
-                tf.cast(
-                    tf.math.greater(
-                        num_valid_images, self.max_images_per_prompt
-                    ),
-                    dtype=tf.int32,
+        # == Branch: vision model, with non-`None` value for `images` ==
+
+        # Check: token IDs should not have less than 0, or more than
+        # `max_images_per_prompt` start of image tokens.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:671]: remove commented code

**Review by divyashreepathihalli — APPROVED** [review]
LGTM, few NIT comments for clean up.

File: keras_hub/src/models/gemma3/gemma3_tokenizer.py [hunk 3]
```diff
@@ -4,6 +4,10 @@
     SentencePieceTokenizer,
 )
 
+START_OF_IMAGE_TOKEN = "<start_of_image>"
+IMAGE_PLACEHOLDER_TOKEN = "<img>"
+END_OF_IMAGE_TOKEN = "<end_of_image>"
```
Comment (abheesht17) [inline keras_hub/src/models/gemma3/gemma3_tokenizer.py:9]: Need to remove this

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 4]
```diff
@@ -339,6 +334,14 @@ def get_config(self):
         )
         return config
 
+    def get_lora_target_names(self):
+        target_names = super().get_lora_target_names()
+
+        # Add these for `Gemma3VITAttention`.
+        if not self.text_only_model:
+            target_names += ["query_proj", "value_proj"]
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_backbone.py:342]: Does it hurt to just always leave these as part of the targets? They just won't match right?

File: keras_hub/src/models/gemma3/gemma3_tokenizer.py (hunk 3 above)
Comment (mattdangerw) [reply keras_hub/src/models/gemma3/gemma3_tokenizer.py:9]: they are no longer used right?

File: keras_hub/src/models/gemma3/gemma3_interleave_embeddings.py [hunk 5]
```diff
@@ -19,45 +19,80 @@ def call(self, image_embeddings, text_embeddings, vision_indices):
...
         Args:
-            image_embeddings: Tensor of shape
-                `(batch_size * num_images_per_prompt,
-                num_vision_tokens_per_image, embedding_dim)`.
-            text_embeddings: Tensor of shape
-            `(batch_size, seq_length, embedding_dim)`.
-            text_mask: Boolean tensor of shape `(batch_size, seq_length)`.
+            image_embeddings: tensor. Image embeddings as returned by the
+                vision encoder (`Gemma3ViT`, usually). Shape:
+            `(batch_size * num_images_per_prompt, num_vision_tokens_per_image,`
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_interleave_embeddings.py:24]: keep this indented so the arg list reads right.

File: keras_hub/src/models/gemma3/gemma3_image_converter.py [hunk 6]
```diff
@@ -6,3 +6,9 @@
 @keras_hub_export("keras_hub.layers.Gemma3ImageConverter")
 class Gemma3ImageConverter(ImageConverter):
     backbone_cls = Gemma3Backbone
+
+    def __init__(self, **kwargs):
+        # Always do image preprocessing in float32
+        kwargs.pop("dtype", None)
+        dtype = "float32"
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_image_converter.py:13]: why is this btw? won't the images get converted to the compute dtype later?

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 7]
```diff
@@ -14,10 +15,6 @@
 from keras_hub.src.utils.tensor_utils import preprocessing_function
 from keras_hub.src.utils.tensor_utils import strip_to_ragged
 
-START_OF_IMAGE_TOKEN = "<start_of_image>"
-IMAGE_PLACEHOLDER_TOKEN = "<img>"
-END_OF_IMAGE_TOKEN = "<end_of_image>"
-
 
 @keras_hub_export("keras_hub.models.Gemma3CausalLMPreprocessor")
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:19]: Not for this PR, but looks like our prompts, responses setup will not work for multi-turn conversations.  We should consider how we want that to work in the future.

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 8]
```diff
@@ -273,162 +371,151 @@ def call(
...
-            if self.image_converter.image_size
-            else original_image_shape[-3]
-        )
-        width = (
-            self.image_size[1]
-            if self.image_converter.image_size
-            else original_image_shape[-2]
-        )
+
+        if keras.config.backend() == "torch" and not isinstance(
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:491]: This seems unsafe, given that in some modes of operations images is user input right? We could also have a np array with a `.cpu()` function for example. Maybe do `images = images.convert_to_numpy(images)` which should handle cpu conversion and other cases.

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 9]
```diff
@@ -448,39 +535,69 @@ def generate_preprocess(
...
+            prompts = [prompts]
+            if responses is not None:
+                responses = [responses]
+        if isinstance(prompts, tf.Tensor) and len(prompts.shape) == 0:
+            batched = False
+            prompts = tf.expand_dims(prompts, axis=0)
+            if responses is not None:
+                responses = tf.expand_dims(responses, axis=0)
+
+        # There are 8 cases, based on values of
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:571]: Consider if there's some common utilities to refactor and share between call/generate. I see a lot of duplicated code.

**Review by mattdangerw — COMMENTED** [review]
Thanks! Haven't done a perfect read through of the preprocessor yet, but left some comments.

File: keras_hub/src/models/gemma3/gemma3_tokenizer.py (hunk 3 above)
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_tokenizer.py:9]: Yeah, no longer used

File: keras_hub/src/models/gemma3/gemma3_image_converter.py (hunk 6 above)
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_image_converter.py:13]: I want to do standardisation, normalisation in `float32` because these ops are sensitive to precision. What do you think?

File: keras_hub/src/models/gemma3/gemma3_backbone.py (hunk 4 above)
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:339]: Hmmm, yeah. But let's keep the if condition for clarity

File: keras_hub/src/models/gemma3/gemma3_decoder_block.py [hunk 10]
```diff
@@ -14,16 +14,17 @@
...
-    - `rope_scaling_factor`: RoPE scaling factor differs from local to global
-      attention layers.
-    - `gate_dim_reduction`: In the gating layers, Gemma and Gemma2 reduce
-      intermediate dimension by 2. For Gemma3, no such reduction happens.
-    - Uses bidirectional attention for images, and causal for everything else.
+    This decoder layer is the same as the layer used for Gemma and Gemma2.
+    However, there are a few key differences. Firstly, image tokens have
+    bidirectional masking. Additionally, this layer exposes the following args:
+
+    `use_query_key_norm`: bool. If True, apply RMS normalization on query
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_decoder_block.py:21]: We don't backtick args generally i think

**Review by mattdangerw — COMMENTED** [review]
Thanks! So misc polish comments as we finish up the t4 errors.

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 11]
```diff
@@ -165,48 +394,152 @@ def _format_output(
...
+        per_sample_bool = tf.math.multiply(per_sample_bool_1, per_sample_bool_2)
+        return tf.cast(tf.math.reduce_prod(per_sample_bool), dtype=bool)
+
+    def _preprocess_images(self, images, batched):
+        # Check: token IDs should not have less than 1, or more than
+        # `max_images_per_prompt` start of image tokens.
+        # def no_op():
+        #     pass
+
+        # def raise_error():
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:460]: This code doesn't work :(

File: keras_hub/src/models/gemma3/gemma3_decoder_block.py (hunk 10 above)
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_decoder_block.py:21]: Yeah, I backticked these, because I did not put them inside `args`. These are just additional args.

File: keras_hub/src/models/gemma3/gemma3_attention.py [hunk 12]
```diff
@@ -275,7 +277,41 @@ def call(
 
             key_update = self._apply_rope(key_update, cache_update_index)
             value_update = self.value_dense(x)
+
+            # Update cache. Note that the cache is updated only if the
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_attention.py:281]: Oh, yeah. Missed this. The expected shape is `bsz, sequence_length` (where sequence_length will be 1 when we generate token-by-token)

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 13]
```diff
@@ -67,29 +64,90 @@ class Gemma3Backbone(Backbone):
...
+    input_data["padding_mask"] = (
+        np.expand_dims(np.array([1] * 286 + [0] * (300 - 286)), axis=0)
+        .astype(bool)
+    )
+
     # Pretrained Gemma3 decoder.
     model = keras_hub.models.Gemma3Backbone.from_preset("gemma3_instruct_4b")
     model(input_data)
 
+    vision_encoder = Gemma3ViT(
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:138]: SD3 has `VAEBackbone`, `CLIPTextEncoder`. So, calling it `SigLIPVisionEncoder` makes more sense?

[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:138]: I'll go with Gemma3VisionEncoder, then

File: keras_hub/src/models/gemma3/gemma3_attention.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_attention.py:281]: Added it to the same comment. `call` does not have a doc-string for now

File: keras_hub/src/models/gemma3/gemma3_backbone.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:138]: Keeping it `Gemma3ViT` for now. It should ideally be `SigLIPVisionEncoder`, but it already exists in the repo, and it'll get confusing that way. Or we can call it `Gemma3SigLIPVisionEncoder`. Mammoth name

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py]: I've removed it for now

[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py]: sg. could just leave a one liner `# TODO: Error when...`

File: keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py [hunk 14]
```diff
@@ -64,25 +69,170 @@ class Gemma3CausalLMPreprocessor(CausalLMPreprocessor):
...
+    # Different number of images in every sample.
+    preprocessor(
+        {
+            "prompts": [
+                "Who is this singer: <start_of_image>?",
+                "Who are these musicians <start_of_image>, <start_of_image>?"
+            ],
+            "responses": ["Arijit Singh", "John Lennon, Paul Mccartney"],
+            "images": [
+                [
```
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_hub/src/models/gemma3/gemma3_causal_lm_preprocessor.py:203]: sgtm!

File: keras_hub/src/models/gemma3/gemma3_backbone.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_hub/src/models/gemma3/gemma3_backbone.py]: `Gemma3VisionEncoder` sounds nice and general, from the transformer point of view it just need to compute soft image tokens right? so is doesn't have to be siglip in there

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 15]
```diff
@@ -108,10 +109,84 @@ class Gemma3Backbone(Backbone):
...
+    input_data["padding_mask"] = (
+        np.expand_dims(np.array([1] * 286 + [0] * (300 - 286)), axis=0)
+        .astype(bool)
+    )
+
+    # Pretrained Gemma3 decoder.
+    model = keras_hub.models.Gemma3Backbone.from_preset("gemma3_instruct_4b")
+    model(input_data)
+
+    vision_encoder = Gemma3VisionEncoder(
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_backbone.py:157]: remove this right? this is the pretrained gemma section?

File: keras_hub/src/models/gemma3/gemma3_backbone.py [hunk 16]
```diff
@@ -108,10 +109,84 @@ class Gemma3Backbone(Backbone):
...
+        hidden_dim=1152,
+        num_layers=27,
+        intermediate_dim=4304,
+        output_dim=2560,
+        pool_size=4,
+        layer_norm_epsilon=1e-6,
+        dtype="float32",
+    )
+
+    model = keras_hub.models.Gemma3Backbone(
```
Comment (mattdangerw) [inline keras_hub/src/models/gemma3/gemma3_backbone.py:170]: and this

**Review by mattdangerw — APPROVED** [review]
Thanks! I'll keep an eye here but dropping an approval in case you want to merge when I am asleep.

File: keras_hub/src/tests/mocks/mock_gemma3_tokenizer.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
+import tensorflow as tf
+
+from keras_hub.src.tokenizers.tokenizer import Tokenizer
+from keras_hub.src.utils.tensor_utils import convert_to_ragged_batch
+from keras_hub.src.utils.tensor_utils import is_int_dtype
+from keras_hub.src.utils.tensor_utils import is_string_dtype
+from keras_hub.src.utils.tensor_utils import preprocessing_function
+
+
+class MockGemma3Tokenizer(Tokenizer):
```
Comment (mattdangerw) [inline keras_hub/src/tests/mocks/mock_gemma3_tokenizer.py:10]: This mock is going to be hard to maintain. What we if just add the real gemma3 spm file to `tests/test_data/gemma3_vocab.spm`? And leave a todo to replace with a smaller test tokenizer?

I'd rather have a slightly oversized file than have to carry this code.

File: keras_hub/src/models/gemma3/gemma3_backbone.py (hunk 15 above)
Comment (abheesht17) [reply keras_hub/src/models/gemma3/gemma3_backbone.py:157]: Don't we show both in doc-strings?
https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/bert/bert_backbone.py#L66C5-L66C62

File: keras_hub/src/tests/mocks/mock_gemma3_tokenizer.py (hunk 17 above)
Comment (abheesht17) [reply keras_hub/src/tests/mocks/mock_gemma3_tokenizer.py:10]: I can move back to the original SPM file, but it's ~4 MB
